File: services/ai/gateway/internal/handler/health.go
package handler

import (
	"encoding/json"
	"net/http"

	"github.com/rs/zerolog"

	"github.com/stocktracker/gateway/internal/cli"
	"github.com/stocktracker/gateway/internal/config"
	"github.com/stocktracker/gateway/internal/db"
)

type HealthHandler struct {
	config   *config.Config
	database *db.PostgresDB
	redis    *db.RedisDB
	cli      *cli.Executor
	logger   zerolog.Logger
}

func NewHealthHandler(cfg *config.Config, database *db.PostgresDB, redis *db.RedisDB, cliExec *cli.Executor, logger zerolog.Logger) *HealthHandler {
	return &HealthHandler{config: cfg, database: database, redis: redis, cli: cliExec, logger: logger}
}

func (h *HealthHandler) Health(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok", "service": "gateway"})
}

func (h *HealthHandler) Live(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

func (h *HealthHandler) Ready(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	checks := map[string]string{}
	allOk := true

	// Check database
	if h.database != nil {
		if err := h.database.HealthCheck(ctx); err != nil {
			checks["database"] = "error: " + err.Error()
			allOk = false
		} else {
			checks["database"] = "ok"
		}
	}

	// Check Redis
	if h.redis != nil {
		if err := h.redis.HealthCheck(ctx); err != nil {
			checks["redis"] = "error: " + err.Error()
			allOk = false
		} else {
			checks["redis"] = "ok"
		}
	}

	// Check CLI availability
	if h.cli != nil {
		if h.cli.CheckCLIAvailable(ctx, "cursor-agent") {
			checks["cursor_agent"] = "ok"
		} else {
			checks["cursor_agent"] = "not available"
			allOk = false
		}
	}

	w.Header().Set("Content-Type", "application/json")
	if !allOk {
		w.WriteHeader(http.StatusServiceUnavailable)
	}

	resp := map[string]interface{}{"status": "ok", "checks": checks}
	if !allOk {
		resp["status"] = "degraded"
	}
	json.NewEncoder(w).Encode(resp)
}

File: services/ai/gateway-2.0/src/http/health.ts
/**
 * Health check route handlers.
 */

import type { FastifyInstance } from 'fastify';
import type { PostgresClient } from '../db/postgres.js';
import type { RedisClient } from '../db/redis.js';
import type { CLIExecutor } from '../core/cli/executor.js';
import type { GatewayConfig } from '../config.js';

export function registerHealthRoutes(
  app: FastifyInstance,
  db: PostgresClient,
  redis: RedisClient,
  cliExecutor: CLIExecutor,
  config: GatewayConfig,
): void {
  // GET /health - basic liveness
  app.get('/health', async (_request, reply) => {
    return reply.send({ status: 'ok', service: 'gateway-2.0' });
  });

  // GET /health/live - Kubernetes liveness probe
  app.get('/health/live', async (_request, reply) => {
    return reply.send({ status: 'ok' });
  });

  // GET /health/ready - readiness probe (checks DB, Redis, CLI)
  app.get('/health/ready', async (_request, reply) => {
    const checks: Record<string, string> = {};
    let allOk = true;

    try {
      await db.healthCheck();
      checks['database'] = 'ok';
    } catch (err) {
      checks['database'] = `error: ${err instanceof Error ? err.message : String(err)}`;
      allOk = false;
    }

    try {
      await redis.healthCheck();
      checks['redis'] = 'ok';
    } catch (err) {
      checks['redis'] = `error: ${err instanceof Error ? err.message : String(err)}`;
      allOk = false;
    }

    try {
      const available = await cliExecutor.checkCLIAvailable();
      checks['cursor_agent'] = available ? 'ok' : 'not available';
      if (!available) allOk = false;
    } catch {
      checks['cursor_agent'] = 'not available';
      allOk = false;
    }

    checks['memory_curator_model'] = config.curatorModel;
    checks['memory_curator_sequential_batches'] = config.curatorSequentialBatches
      ? 'on'
      : 'off';
    checks['memory_curator_llm_timeout_ms'] = String(config.curatorLlmTimeoutMs);
    checks['memory_curator_max_stories'] = String(config.curatorMaxStories);
    checks['memory_curator_max_stories_per_batch'] = String(
      config.curatorMaxStoriesPerBatch,
    );
    checks['cursor_api_key_configured'] = config.cursorApiKeyConfigured ? 'yes' : 'no';

    const status = allOk ? 'ok' : 'degraded';
    const statusCode = allOk ? 200 : 503;
    return reply.status(statusCode).send({ status, checks });
  });
}
